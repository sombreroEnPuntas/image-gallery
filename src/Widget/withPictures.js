import React, { Component } from 'react'

const withPictures = WrappedComponent =>
  class WithPictures extends Component {
    state = {
      pictures: {},
    }

    handleAddPicture = index => ({ file, image }) =>
      !!image &&
      !!file &&
      file.type.includes('image/') &&
      this.setState(prevState => {
        let newState = Object.assign({}, prevState)
        newState.pictures[index] = image
        return newState
      })

    handleRemovePicture = index => () =>
      this.setState(prevState => {
        let newState = Object.assign({}, prevState)
        delete newState.pictures[index]
        return newState
      })

    render() {
      const { ...passThroughProps } = this.props
      const { pictures } = this.state

      const newProps = {
        pictures,
        addPicture: this.handleAddPicture,
        removePicture: this.handleRemovePicture,
      }

      return <WrappedComponent {...passThroughProps} {...newProps} />
    }
  }

export default withPictures
