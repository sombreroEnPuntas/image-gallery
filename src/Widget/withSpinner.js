import React, { Component } from 'react'
import PropTypes from 'prop-types'

const totalCards = 5

const withSpinner = WrappedComponent =>
  class WithSpinner extends Component {
    static propTypes = {
      initialCard: PropTypes.number.isRequired,
    }

    static defaultProps = {
      initialCard: 0,
    }

    state = {
      currentCard: this.props.initialCard,
    }

    handleClickBack = () => {
      this.setState(prevState => ({
        currentCard:
          prevState.currentCard > 0
            ? prevState.currentCard - 1
            : totalCards - 1,
      }))
    }

    handleClickNext = () => {
      this.setState(prevState => ({
        currentCard:
          prevState.currentCard < totalCards - 1
            ? prevState.currentCard + 1
            : 0,
      }))
    }

    render() {
      const { ...passThroughProps } = this.props
      const { currentCard } = this.state

      const newProps = {
        currentCard,
        onClickBack: this.handleClickBack,
        onClickNext: this.handleClickNext,
        totalCards,
      }

      return <WrappedComponent {...passThroughProps} {...newProps} />
    }
  }

export default withSpinner
