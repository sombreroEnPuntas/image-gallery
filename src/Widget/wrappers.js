import styled from 'styled-components'

export const Container = styled.div`
  align-items: center;
  display: flex;
  justify-content: center;
`

export const Cards = styled.div`
  display: flex;
  flex-flow: row nowrap;
  height: 220px;
  justify-content: center;
  width: 50%;
`

export const Position = styled.span`
  border: ${props =>
    props.active ? '2px solid black' : '2px solid transparent'};
  display: inline-block;
  height: 150px;
  width: 150px;

  &:nth-child(1) {
    transform: rotate(-40deg) scale(0.7) translate(60%, 120%);
    z-index: 0;
  }
  &:nth-child(2) {
    transform: rotate(-20deg) scale(0.9) translate(35%, 25%);
    z-index: 1;
  }
  &:nth-child(3) {
    z-index: 2;
  }
  &:nth-last-child(2) {
    transform: rotate(20deg) scale(0.9) translate(-35%, 25%);
    z-index: 1;
  }
  &:nth-last-child(1) {
    transform: rotate(40deg) scale(0.7) translate(-60%, 120%);
    z-index: 0;
  }
`

export const Image = styled.img`
  background-color: #1cb3ff;
  height: 150px;
  width: 150px;
`

export const Button = styled.button`
  border-radius: 50%;
  border: 0;
  height: 32px;
  padding: 0;
  width: 32px;
  z-index: 3;
`
