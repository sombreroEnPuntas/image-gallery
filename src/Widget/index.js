import React, { Component } from 'react'
import PropTypes from 'prop-types'

import withSpinner from './withSpinner'

// Styled
import { Position, Cards, Container, Button, Image } from './wrappers'

const totalCards = 5

export class Widget extends Component {
  render() {
    const { pictures, currentCard, onClickBack, onClickNext } = this.props

    if (pictures.length < 1) {
      return <div>Loading...</div>
    }

    const cards = []
    let index = 0

    for (; index <= totalCards - 1; index++) {
      cards.push(
        <Position key={`card-${index}`} active={currentCard === index}>
          <Image src={pictures[index].src} alt={pictures[index].alt} />
        </Position>
      )
    }

    return (
      <Container>
        <Button onClick={onClickBack}>&lt;</Button>

        <Cards>{cards}</Cards>

        <Button onClick={onClickNext}>&gt;</Button>
      </Container>
    )
  }
}

Widget.propTypes = {
  pictures: PropTypes.arrayOf(
    PropTypes.shape({ alt: PropTypes.string, src: PropTypes.string })
  ),
  currentCard: PropTypes.number.isRequired,
  onClickBack: PropTypes.func.isRequired,
  onClickNext: PropTypes.func.isRequired,
}

Widget.defaultProps = {
  reviews: [],
}

export default withSpinner(Widget)
