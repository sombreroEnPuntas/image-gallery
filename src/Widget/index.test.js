import React from 'react'
import { mount } from 'enzyme'

import Widget from './'

// Mocks
import mockPictures from '../mocks/pictures'

describe('Widget renders mocks as expected for:', () => {
  it('Loading message...', () => {
    const wrapper = mount(<Widget pictures={[]} totalCards={0} />)
    expect(wrapper.debug()).toMatchSnapshot()
  })

  it('Pictures!', () => {
    const wrapper = mount(
      <Widget pictures={mockPictures} totalCards={mockPictures.length} />
    )
    expect(wrapper.debug()).toMatchSnapshot()
  })
})
