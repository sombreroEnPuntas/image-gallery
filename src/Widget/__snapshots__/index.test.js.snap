// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget renders mocks as expected for: Loading message... 1`] = `
"<WithSpinner pictures={{...}} totalCards={0} initialCard={0}>
  <Widget pictures={{...}} totalCards={0} initialCard={0} currentCard={0} onClickBack={[Function]} onClickNext={[Function]} reviews={{...}}>
    <div>
      Loading...
    </div>
  </Widget>
</WithSpinner>"
`;

exports[`Widget renders mocks as expected for: Pictures! 1`] = `
"<WithSpinner pictures={{...}} totalCards={5} initialCard={0}>
  <Widget pictures={{...}} totalCards={5} initialCard={0} currentCard={0} onClickBack={[Function]} onClickNext={[Function]} reviews={{...}}>
    <styled.div>
      <StyledComponent forwardedComponent={{...}} forwardedRef={{...}}>
        <div className=\\"sc-bdVaJa emmSvS\\">
          <styled.button onClick={[Function]}>
            <StyledComponent onClick={[Function]} forwardedComponent={{...}} forwardedRef={{...}}>
              <button onClick={[Function]} className=\\"sc-ifAKCX gVAEIN\\">
                &lt;
              </button>
            </StyledComponent>
          </styled.button>
          <styled.div>
            <StyledComponent forwardedComponent={{...}} forwardedRef={{...}}>
              <div className=\\"sc-bwzfXH kxWXJw\\">
                <styled.span active={true}>
                  <StyledComponent active={true} forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-htpNat gguXpr\\">
                      <styled.img src=\\"https://source.unsplash.com/random/150x150?sig=0\\" alt=\\"some description 0\\">
                        <StyledComponent src=\\"https://source.unsplash.com/random/150x150?sig=0\\" alt=\\"some description 0\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <img src=\\"https://source.unsplash.com/random/150x150?sig=0\\" alt=\\"some description 0\\" className=\\"sc-bxivhb jElvlS\\" />
                        </StyledComponent>
                      </styled.img>
                    </span>
                  </StyledComponent>
                </styled.span>
                <styled.span active={false}>
                  <StyledComponent active={false} forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-htpNat joeKyc\\">
                      <styled.img src=\\"https://source.unsplash.com/random/150x150?sig=1\\" alt=\\"some description 1\\">
                        <StyledComponent src=\\"https://source.unsplash.com/random/150x150?sig=1\\" alt=\\"some description 1\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <img src=\\"https://source.unsplash.com/random/150x150?sig=1\\" alt=\\"some description 1\\" className=\\"sc-bxivhb jElvlS\\" />
                        </StyledComponent>
                      </styled.img>
                    </span>
                  </StyledComponent>
                </styled.span>
                <styled.span active={false}>
                  <StyledComponent active={false} forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-htpNat joeKyc\\">
                      <styled.img src=\\"https://source.unsplash.com/random/150x150?sig=2\\" alt=\\"some description 2\\">
                        <StyledComponent src=\\"https://source.unsplash.com/random/150x150?sig=2\\" alt=\\"some description 2\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <img src=\\"https://source.unsplash.com/random/150x150?sig=2\\" alt=\\"some description 2\\" className=\\"sc-bxivhb jElvlS\\" />
                        </StyledComponent>
                      </styled.img>
                    </span>
                  </StyledComponent>
                </styled.span>
                <styled.span active={false}>
                  <StyledComponent active={false} forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-htpNat joeKyc\\">
                      <styled.img src=\\"https://source.unsplash.com/random/150x150?sig=3\\" alt=\\"some description 3\\">
                        <StyledComponent src=\\"https://source.unsplash.com/random/150x150?sig=3\\" alt=\\"some description 3\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <img src=\\"https://source.unsplash.com/random/150x150?sig=3\\" alt=\\"some description 3\\" className=\\"sc-bxivhb jElvlS\\" />
                        </StyledComponent>
                      </styled.img>
                    </span>
                  </StyledComponent>
                </styled.span>
                <styled.span active={false}>
                  <StyledComponent active={false} forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-htpNat joeKyc\\">
                      <styled.img src=\\"https://source.unsplash.com/random/150x150?sig=4\\" alt=\\"some description 4\\">
                        <StyledComponent src=\\"https://source.unsplash.com/random/150x150?sig=4\\" alt=\\"some description 4\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <img src=\\"https://source.unsplash.com/random/150x150?sig=4\\" alt=\\"some description 4\\" className=\\"sc-bxivhb jElvlS\\" />
                        </StyledComponent>
                      </styled.img>
                    </span>
                  </StyledComponent>
                </styled.span>
              </div>
            </StyledComponent>
          </styled.div>
          <styled.button onClick={[Function]}>
            <StyledComponent onClick={[Function]} forwardedComponent={{...}} forwardedRef={{...}}>
              <button onClick={[Function]} className=\\"sc-ifAKCX gVAEIN\\">
                &gt;
              </button>
            </StyledComponent>
          </styled.button>
        </div>
      </StyledComponent>
    </styled.div>
  </Widget>
</WithSpinner>"
`;
